pandas/core/algorithms.py

```diff
@@ -169,22 +173,24 @@ def _ensure_data(values, dtype=None):
 
     # we have failed, return object
     values = np.asarray(values, dtype=np.object)
-    return ensure_object(values), "object"
+    return ensure_object(values), np.dtype("object")
 
 
-def _reconstruct_data(values, dtype, original):
+def _reconstruct_data(
+    values: ArrayLike, dtype: DtypeObj, original: AnyArrayLike
+) -> ArrayLike:
```

Comment:
ArrayLike is a typevar. is the return type here always the same as the type of values? or dependant on dtype. 

maybe just expand on the prose in the docstring till this bites us.

Reply:
> ArrayLike is a typevar. is the return type here always the same as the type of values? or dependant on dtype.

Yikes, I think I've been using it incorrectly in a lot of places then.  Ive been using it as a synonym for `Union[np.ndarray, ExtensionArray]`.  I guess I'll do a dedicated pass through the code to weed out those mis-usages

Reply:
> Ive been using it as a synonym for `Union[np.ndarray, ExtensionArray]`

This is OK if the alias only appears once in the function signature/return.

once it appears twice, then they are bound.